File: lib/go/corpus_test.go
package gencmu

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"reflect"
	"runtime"
	"sort"
	"strconv"
	"strings"
	"sync"
	"testing"
	"time"
)

// The shared Lojban corpus (tests/README.md, "Corpus cases"): the core
// sample by default, every case with GENCMU_CORPUS=full. Cases run on
// GENCMU_CORPUS_WORKERS goroutines (the number of CPUs by default), which
// share one *Dialect per dialect.

type corpusCase struct {
	ID       string
	Text     string
	Dialect  string
	Features []string
	fields   map[string]any // the fields compared
}

var corpusFields = []string{"expect", "verdict", "stage", "ties", "words", "brackets"}

func readCorpus(t *testing.T) []*corpusCase {
	files, _ := filepath.Glob("../../tests/corpus/*.jsonl")
	sort.Strings(files)
	if len(files) == 0 {
		t.Fatal("no corpus in ../../tests/corpus")
	}
	var cases []*corpusCase
	for _, f := range files {
		fh, err := os.Open(f)
		if err != nil {
			t.Fatal(err)
		}
		sc := bufio.NewScanner(fh)
		sc.Buffer(make([]byte, 1<<20), 1<<26)
		for sc.Scan() {
			line := strings.TrimSpace(sc.Text())
			if line == "" {
				continue
			}
			c := &corpusCase{}
			if err := json.Unmarshal([]byte(line), c); err != nil {
				t.Fatalf("%s: %v", f, err)
			}
			var all map[string]any
			json.Unmarshal([]byte(line), &all)
			c.fields = map[string]any{}
			for _, k := range corpusFields {
				if v, ok := all[k]; ok {
					c.fields[k] = v
				}
			}
			cases = append(cases, c)
		}
		fh.Close()
		if err := sc.Err(); err != nil {
			t.Fatalf("%s: %v", f, err)
		}
	}
	return cases
}

// corpusOutcome is what gencmu makes of a case, in the case's own terms.
func corpusOutcome(d *Dialect, c *corpusCase) (map[string]any, error) {
	res, err := d.Parse(c.Text, ParseOptions{Features: c.Features})
	if err != nil {
		return nil, err
	}
	got := map[string]any{}
	if res.OK {
		got["expect"] = "accept"
		got["verdict"] = res.Stages[len(res.Stages)-1].Verdict
		got["brackets"] = Brackets(res, BracketOptions{})
	} else {
		got["expect"] = "reject"
		if res.Error != nil && res.Error.Stage != "" {
			got["stage"] = res.Error.Stage
		} else {
			got["stage"] = nil
		}
	}
	var ties []any
	for _, s := range res.Stages {
		if s.Verdict == VerdictTie {
			ties = append(ties, s.Name)
		}
		if s.Name == "words" && s.Output != nil {
			words := make([]any, len(s.Output))
			for i, tok := range s.Output {
				if tok.Phonemes != "" {
					words[i] = tok.Phonemes
				} else {
					words[i] = tok.Text
				}
			}
			got["words"] = words
		}
	}
	if len(ties) > 0 {
		got["ties"] = ties
	}
	return got, nil
}

func TestCorpus(t *testing.T) {
	cases := readCorpus(t)
	full := os.Getenv("GENCMU_CORPUS") == "full"
	if !full {
		data, err := os.ReadFile("../../tests/core.txt")
		if err != nil {
			t.Fatal(err)
		}
		core := map[string]bool{}
		for _, id := range strings.Fields(string(data)) {
			core[id] = true
		}
		var sample []*corpusCase
		for _, c := range cases {
			if core[c.ID] {
				sample = append(sample, c)
			}
		}
		if len(sample) != len(core) {
			t.Fatalf("core.txt names %d ids, of which %d are cases", len(core), len(sample))
		}
		cases = sample
	}
	dialects := map[string]*Dialect{}
	for _, c := range cases {
		if dialects[c.Dialect] == nil {
			d, err := LoadDialect(c.Dialect)
			if err != nil {
				t.Fatalf("%s: %v", c.Dialect, err)
			}
			dialects[c.Dialect] = d
		}
	}
	// The longest texts first, so that the pool is not left waiting on one.
	queue := append([]*corpusCase(nil), cases...)
	sort.SliceStable(queue, func(i, j int) bool { return len(queue[i].Text) > len(queue[j].Text) })
	workers := runtime.NumCPU()
	if s := os.Getenv("GENCMU_CORPUS_WORKERS"); s != "" {
		workers, _ = strconv.Atoi(s)
	}
	started := time.Now()
	jobs := make(chan *corpusCase)
	var mu sync.Mutex
	var failures []string
	var wg sync.WaitGroup
	for w := 0; w < max(1, workers); w++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for c := range jobs {
				failure := checkCorpusCase(dialects[c.Dialect], c)
				if failure != "" {
					mu.Lock()
					failures = append(failures, failure)
					mu.Unlock()
				}
			}
		}()
	}
	for _, c := range queue {
		jobs <- c
	}
	close(jobs)
	wg.Wait()
	var mem runtime.MemStats
	runtime.ReadMemStats(&mem)
	t.Logf("%d cases on %d goroutines in %v; %d differ; heap obtained from the OS %d MB", len(cases), workers, time.Since(started).Round(time.Millisecond), len(failures), mem.Sys>>20)
	sort.Strings(failures)
	for i, f := range failures {
		if i == 20 {
			t.Errorf("... and %d more", len(failures)-20)
			break
		}
		t.Error(f)
	}
}

func checkCorpusCase(d *Dialect, c *corpusCase) (failure string) {
	defer func() {
		if x := recover(); x != nil {
			failure = fmt.Sprintf("%s: panic: %v", c.ID, x)
		}
	}()
	got, err := corpusOutcome(d, c)
	if err != nil {
		return fmt.Sprintf("%s: %v", c.ID, err)
	}
	for _, k := range corpusFields {
		want, inCase := c.fields[k]
		have, inGot := got[k]
		if !inCase && !inGot {
			continue
		}
		if !reflect.DeepEqual(want, have) {
			w, _ := json.Marshal(want)
			h, _ := json.Marshal(have)
			if !inCase {
				w = []byte("(absent)")
			}
			if !inGot {
				h = []byte("(absent)")
			}
			return fmt.Sprintf("%s (%s): %s expected %s, got %s", c.ID, c.Dialect, k, w, h)
		}
	}
	return ""
}

File: lib/js/test/corpus.test.js
// Runs the shared Lojban corpus (tests/README.md, "Corpus cases"): the core
// sample by default, every case with GENCMU_CORPUS=full. Cases run on a pool
// of worker threads, each loading the bundled dialects once.
import { test } from "node:test";
import assert from "node:assert/strict";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { Worker, isMainThread, parentPort } from "node:worker_threads";

const here = path.dirname(fileURLToPath(import.meta.url));
const tests = path.join(here, "..", "..", "..", "tests");

/** Every corpus case, in file order. */
function allCases() {
  const cases = [];
  const directory = path.join(tests, "corpus");
  for (const file of fs.readdirSync(directory).filter((name) => name.endsWith(".jsonl")).sort()) {
    for (const line of fs.readFileSync(path.join(directory, file), "utf8").split("\n")) {
      if (line.trim()) cases.push(JSON.parse(line));
    }
  }
  return cases;
}

/** What gencmu produces for a case, in the case's own terms. */
async function outcome(g, dialects, c) {
  if (!dialects.has(c.dialect)) dialects.set(c.dialect, g.loadDialect(c.dialect));
  const result = dialects.get(c.dialect).parse(c.text, { features: c.features || [] });
  const words = result.stages.find((stage) => stage.name === "words");
  const got = { expect: result.ok ? "accept" : "reject" };
  if (result.ok) got.verdict = result.stages[result.stages.length - 1].verdict;
  else got.stage = result.error ? result.error.stage : null;
  const ties = result.stages.filter((stage) => stage.verdict === "tie").map((stage) => stage.name);
  if (ties.length) got.ties = ties;
  if (words && words.output) got.words = words.output.map((token) => token.phonemes || token.text);
  if (result.ok) got.brackets = g.toBrackets(result);
  return got;
}

if (!isMainThread) {
  const g = await import("../src/node.js");
  const dialects = new Map();
  parentPort.on("message", async (c) => {
    try {
      parentPort.postMessage({ id: c.id, got: await outcome(g, dialects, c) });
    } catch (error) {
      parentPort.postMessage({ id: c.id, crash: String(error && error.stack || error) });
    }
  });
} else {
  test("the corpus", async () => {
    const full = process.env.GENCMU_CORPUS === "full";
    let cases = allCases();
    if (!full) {
      const core = new Set(fs.readFileSync(path.join(tests, "core.txt"), "utf8").split("\n").filter(Boolean));
      cases = cases.filter((c) => core.has(c.id));
      assert.equal(cases.length, core.size, "every id of core.txt names a case");
    }
    const byId = new Map(cases.map((c) => [c.id, c]));
    // The longest texts first, so that the pool is not left waiting on one.
    const queue = cases.slice().sort((a, b) => b.text.length - a.text.length);
    const failures = [];
    const workers = Math.max(1, Math.min(queue.length, Number(process.env.GENCMU_CORPUS_WORKERS) || os.availableParallelism() - 1 || 1));
    await Promise.all(Array.from({ length: workers }, () => new Promise((resolve, reject) => {
      // A chapter of a book takes over a gigabyte to parse, which may be
      // more than a worker's default heap on a small machine.
      const worker = new Worker(fileURLToPath(import.meta.url), { resourceLimits: { maxOldGenerationSizeMb: 3072 } });
      const feed = () => {
        const next = queue.pop();
        if (next) worker.postMessage(next);
        else worker.terminate().then(resolve, reject);
      };
      worker.on("message", (message) => {
        const c = byId.get(message.id);
        if (message.crash) failures.push(`${c.id}: ${message.crash}`);
        else {
          for (const key of ["expect", "verdict", "stage", "ties", "words", "brackets"]) {
            if (!(key in c) && !(key in message.got)) continue;
            if (JSON.stringify(c[key]) !== JSON.stringify(message.got[key])) {
              failures.push(`${c.id} (${c.dialect}): ${key} expected ${JSON.stringify(c[key])}, got ${JSON.stringify(message.got[key])}`);
              break;
            }
          }
        }
        feed();
      });
      worker.on("error", reject);
      feed();
    })));
    assert.deepEqual(failures.slice(0, 20), [], `${failures.length} of ${cases.length} cases differ`);
  });
}
